### PR TITLE
Mini aod substructure pieces overhaul

### PR DESCRIPTION
**Pull request by rappoccio**
Making new PR to accomplish the goals in  #7606, after code review. The only major change is to write out ONLY subjets that are associated to selected jets. 

====== EVENT Content =======
The AK8 jet content is therefore : 
- Add all AK8 jets for pt > 50 GeV
  - If AK8 jet pt > 200:
    - Add top-tagging and soft drop tagger info. 
    - Run b-tagging and JECs on subjets

So if the AK8 jet has pt between 50 and 200 GeV, the substructure algorithms are not run. If pt > 200 GeV, The subjets are written along with the "fat" jets in the BoostedJetMerger, and then are rekeyed accordingly with the JetSubstructurePacker. The slimming is done upstream on the pat::Jets made for the substructure algorithms, and so downstream this is no longer necessary to re-slim. 

The jet content is (on 9000 events without pileup):

patJets_slimmedJets__PAT    1.00    1.00    1.00        4.8%
patJets_packedPatJetsAK8__PAT   1.00    0.58    0.58        2.8%
recoGenJets_slimmedGenJets__PAT 1.00    0.46    0.46        2.2%
recoGenJets_slimmedGenJetsAK8__PAT  1.00    0.45    0.45        2.2%
patJets_slimmedJetsCMSTopTagCHSPacked_SubJets_PAT   1.00    0.16    0.16        0.8%
patJets_slimmedJetsAK8PFCHSSoftDropPacked_SubJets_PAT   1.00    0.14    0.14        0.7%

So the additional collections (the subjets) are about 1.5% of the total miniAOD size. 

====== CPU usage =====
The additional CPU usage here is negligible because the workflow is not running any algorithms, but instead taking the values directly from the RECO collections upstream and rekeying the information : 

TimeReport   0.000074     0.000074     0.000074  caTopTagInfos
TimeReport   0.000041     0.000041     0.000041  caTopTagInfosPAT
TimeReport   0.000100     0.000100     0.000100  slimmedGenJets
TimeReport   0.000128     0.000128     0.000128  slimmedGenJetsAK8
TimeReport   0.000131     0.000131     0.000131  slimmedJets
TimeReport   0.000000     0.000000     0.000000  slimmedJetsAK8
TimeReport   0.000041     0.000041     0.000041  slimmedJetsAK8PFCHSSoftDropPacked
TimeReport   0.000051     0.000051     0.000051  slimmedJetsAK8PFCHSSoftDropSubjets
TimeReport   0.000048     0.000048     0.000048  slimmedJetsCMSTopTagCHSPacked
TimeReport   0.000056     0.000056     0.000056  slimmedJetsCMSTopTagCHSSubjets
TimeReport   0.000095     0.000095     0.000095  packedPatJetsAK8

==== Testing ====
Tests are checking out (thanks to @ianna  for providing the correct recipe so quickly, I've merged in that PR, and apologies to @ferencek since I had to revert out the genJetsNoNu for the b-tagging sequence, to be discussed at an upcoming JMAR meeting). 

==== For posterity, here is the original PR message : ====

With @arizzi and @ferencek : we've added substructure into miniAOD in a slightly more natural way to allow for multiple interpretations of the jet. This will now be accessed through the "subjet" methods. They can take string or integer access similar to the JECs. I have also adjusted pt thresholds so that they are sensible given common user preferences. This makes the total of the AK8 jets (including two subjet collections) 3.8% of the total miniAOD size.

The nice additions are :

Subjets from soft drop and the CMS top tagger are stored as lists of pat::Jets.
Jet corrections are applied to subjets.
Subjet b-tagging is being run.

### CODE REVIEW

**Comment by ferencek** [comment]
Looks complete to me. The only missing piece, as far as I can tell, is the slimming of `packedPatJetsAK8`. For this it should be sufficient to replace `selectedPatJetsAK8` with `packedPatJetsAK8` in https://github.com/rappoccio/cmssw/blob/MiniAODSubstructurePiecesOverhaul/PhysicsTools/PatAlgos/python/slimming/slimmedJets_cfi.py and update accordingly the event content in https://github.com/rappoccio/cmssw/blob/MiniAODSubstructurePiecesOverhaul/PhysicsTools/PatAlgos/python/slimming/slimming_cff.py

File: PhysicsTools/PatAlgos/python/slimming/applySubstructure_cfi.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    # add Njetiness
+    process.load('RecoJets.JetProducers.nJettinessAdder_cfi')
+    process.NjettinessAK8 = process.Njettiness.clone()
+    process.NjettinessAK8.src = cms.InputTag("ak8PFJetsCHS")
+    process.NjettinessAK8.cone = cms.double(0.8)
+    process.patJetsAK8.userData.userFloats.src += ['NjettinessAK8:tau1','NjettinessAK8:tau2','NjettinessAK8:tau3']
+
+
+    ## PATify pruned fat jets
```
Comment (ferencek) [inline PhysicsTools/PatAlgos/python/slimming/applySubstructure_cfi.py]: Minor detail: pruned -> soft drop in the comment

**Comment by arizzi** [comment]
Please check the comparison here:
http://arizzi.web.cern.ch/arizzi/subjet.html vs http://arizzi.web.cern.ch/arizzi/refib.html

it looks like patJets_slimmedJetsAK8__PAT grew from 1.6 items per event to 4 items per event.
What happened?

**Comment by rappoccio** [comment]
@arizzi : I reduced the pt cut on AK8 jets WITHOUT substructure from 200 GeV to 50 GeV. Only jets with pt > 200 GeV have substructure attached (including subjets). 

@ferencek : The slimming of that collection won't impact anything, the pt cut for the AK8 jets is derived from [1] and [2], and then associated at [3]. Thus, it picks up the pt cut from the AK8 jets (pt > 50 GeV). 

[1] https://github.com/rappoccio/cmssw/blob/MiniAODSubstructurePiecesOverhaul/PhysicsTools/PatAlgos/python/slimming/applySubstructure_cfi.py#L26
[2] https://github.com/rappoccio/cmssw/blob/MiniAODSubstructurePiecesOverhaul/PhysicsTools/PatAlgos/python/slimming/applySubstructure_cfi.py#L78
https://github.com/rappoccio/cmssw/blob/MiniAODSubstructurePiecesOverhaul/PhysicsTools/PatAlgos/python/slimming/applySubstructure_cfi.py#L110
[3] https://github.com/rappoccio/cmssw/blob/MiniAODSubstructurePiecesOverhaul/PhysicsTools/PatAlgos/python/slimming/applySubstructure_cfi.py#L145

**Comment by arizzi** [comment]
@rappoccio I would then at least slim the daughters of AK8. Anyhow for those having subjets the daughters should be accessible via the subjets (or are there orphanes?) and for the others as we do not store the subjets we may well not store the daughters.

**Comment by rappoccio** [comment]
@arizzi : I'm not sure I follow. The daughters of the AK8 jets are slimmed already (they're just pointers to the packed candidates). The subjets are accessible via the new "subjets" method in pat::Jet, and those in turn are pat::Jets that also have constituents pointing to the packed candidates. The subjets are slimmed (I implemented that yesterday at your suggestion) so that only the relevant ones are kept (i.e. pointed to by a jet that passes the pt cuts).

**Comment by ferencek** [comment]
@rappoccio I think there is some misunderstanding on the meaning of "slimming" in this case. I am not talking about applying a pt cut. I am talking about the MiniAOD slimming of PAT jets where unnecessary data members are zeroed. Can you just implement the changes I suggested above an re-run? The size of `slimmedJetAK8` should reduce.

**Comment by rappoccio** [comment]
@ferencek : Ahhhh, okay, I think I understand what you mean now. I'll put it in.

**Comment by rappoccio** [comment]
@ferencek : OK I tried this but it didn't affect the size at all (over 9k events). That basically makes sense because what is kept is the "slimmedPatJetsAK8" (already slimmed) and then I add the substructure jets as ptrs. The slimming of "slimmedPatJetsAK8" is also inherited by "packedPatJetsAK8". 

Other than that I think we're good to go, no?

**Comment by ferencek** [comment]
@rappoccio There is no collection called `slimmedPatJetsAK8` so I'm not sure to which collection you are referring. I did some quick tests on 200 events only and slimming of `packedPatJetsAK8` resulted in a reduction of ~0.25% on the total event size. In any case, running `packedPatJetsAK8` through the PATJetSlimmer is necessary to get the daughter pointers rekeyed.

Possible ways to save space would be to raise the AK8 threshold from 50 GeV to something in-between 50 and 200 GeV (or just 200 GeV?). Also, we probably don't need all Ak8 GenJets but only those above some threshold. Other possibilities are those Andrea mentioned, namely zeroing daughter pointers for some AK8 jets. The problem is that subjet daughters will not contain all daughters of AK8 jets since subjets come out of a jet grooming/top tagging procedure which filters out some constituents.

**Comment by arizzi** [comment]
@ferencek please also consider clearing the daugther vectors (there is a switch in the slimmer configuration) if the information is not needed

**Comment by rappoccio** [comment]
Please hold off on testing so we can hash out the details. I have added a commit from @ferencek with a few additional slimming pieces. @arizzi @ferencek and I are testing.

**Comment by ferencek** [comment]
What is the final verdict here? Just to remind, tests done on 9k ttbar events with pileup indicate that this now gives us additional ~0.24 kb/event. Is this acceptable?

**Comment by rappoccio** [comment]
We're waiting on a political appointment of someone to make the decision ;).

**Comment by gpetruc** [comment]
Hi Sal, 

Now with this PR you save the daughters of the subjects but not of the fat jet, but if I have understood correctly some of the daughters of the original fat jet might not end up in any of the subjets, so people who want to do subjets with another algorithm won't be able to do it.

I would propose then that when you add the subjets to the fat jet, you also modify the daughters list so that it contains the subjets (for one of the algorithms) and the extra daughters that were in the fat jet but didn't end up in any of the subjets made from that algorithm.

That way, people who want to recluster with a different approach can still get the full list of constituents just by gettting all descendents of the fat jet (all grand-daughters + all daughters), and the size increase should be tiny since most of the daughters are already in the subjets.

**Comment by rappoccio** [comment]
Hi Giovanni,

The extra daughters are kept already. The mechanism here is to collate the "subjet" 4-vectors in this additional "subjets" field in pat::Jet, which also then point to their individual daughters. However, the "pat::Jet::daughters" methods are the same. To get the constituents of the subjets, one would do "pat::Jet::subjets::daughters". This is similar to the "pieces" mechanism of fastjet::PseudoJet.

**Comment by ferencek** [comment]
@rappoccio I went through the code changes again to remind myself what is done and Giovanni is right, daughters of AK8 jets are dropped. However, keeping all of them would be redundant so just storing extra constituents not already contained in subjets should be sufficient. This, however, implies updating the jet slimmer module since at the moment it either keeps or drops all the daughters.

**Comment by rappoccio** [comment]
@ferencek I thought ALL packed candidates were stored, whether or not they are in a jet, right? The AK8 jets are just pointing to "packedCandidates". 

Unless I was mistaken, and only preclustered AK4 packed candidates are stored?

**Comment by ferencek** [comment]
@rappoccio Yes, all PF candidates are stored as packedCandidates. However, in this PR links to daughters of AK8 jets are dropped. Now, one can (almost) get them by first going to the subjets but subjets will not necessarily contain all constituents of the original AK8 jet.

**Comment by rappoccio** [comment]
@ferencek Ooooo, you're right. I'd say we probably need to keep the daughters for the full AK8 jets in that case, which will increase the size a bit.

**Comment by arizzi** [comment]
@rappoccio no you cannot keep the daughters as it would increase too much
the size per event, the solution me and giovanni propose is to do what said
in giovanni's post

@ferencek please update the "Packer" to have the mixed collection, then in
the slimmer we may add an option to avoid rekeying or at least "leave as
is" if the key is not found in the translation valuemap.

**Comment by rappoccio** [comment]
@arizzi I'm not sure this is all that helpful, though. If we add two packed candidate collections off miniaod, it's actually less annoying to just recluster all of the packed candidates directly. People would have to loop over the subjets, get the list of daughters, and then get the "missing" ones from another collection. Grrr. Furthermore, what is "missing" is only defined per boosted jet algorithm. So which "missing" ones are we going to add? Missed by pruning? Soft drop? CMS tagger? It gets kind of more annoying than it's worth to bother. 

IMO we can just leave it as is and catch the requisite 80% of use cases.

**Comment by arizzi** [comment]
@rappoccio there is nothing ambiuous: the daughters will contain

- the  subjets of a given collection (just choose one, it is irrelevant
which one)
- the missing particles (relative to the chosen subet collection)

what ever choice one does of the  subjets to use (let say Soft drop), the
missing packedcands will be the one not catched by those subets.

The fact that user get annoyed is irrelevant, they have all the information
without the need to run a CPU intensive task.
What's the big deal with something like?

constituents=[]
for c in jet.daughterPtrVector() :
    if c.numberOfDaughters() > 0 :
        constituents.extend(c.daughterPtrVector())
    else :
        constrituents.append(c)


we can even make it a standard accessor of CompositeCandidatePtr, something
like
expandedDaughterPtrVector(int maxDepth)

**Comment by rappoccio** [comment]
@arizzi I know it's possible and how to implement it, I just doubt anyone will bother when it's less work to just start from all the packed candidates. For the expert person who will do something nonstandard, I think they're going to be using something industrial-strength anyway so will not blink about just reclustering with AK8. 

In any case it's something we can add later on. Let's get this PR in and see what to do from there. This sounds like a bit of development for almost no users to use it.

**Comment by gpetruc** [comment]
Hi Sal,

No, reclustering the entire event is definitely more cpu consuming than
reclustering only the constituents of a fat jet, and that was possible in
previous versions of miniAOD, i don't see why we should take in for 740 a
PR that breaks such functionality when it can be instead preserved with
only a minimal extra effort.

If the argument is 'well, people can always go back to the full
pfcandidates', that applies as well for the information being added here.

Ciao,

    Giovanni
Il 9-mar-2015 22:27 "rappoccio" <notifications@github.com> ha scritto:

**Comment by rappoccio** [comment]
@gpetruc It's basically a schema shift either way. The number of people who will recluster AFTER this pull request will be much fewer than those before because this is the first time subjet b-tagging is usable, and basically everyone except W/Z tagging needs it. 

Nevertheless, if you guys desperately want this, I can try to implement it tomorrow (although I think it's better in pat::Jet than in CompositeCandidatePtr).

**Comment by rappoccio** [comment]
Please hold off on the tests here, I'm just rebasing.

**Comment by rappoccio** [comment]
Grr, looks like git went haywire with the rebase attempt. It added all of my commits as fresh ones with new ID's after I replayed them on CMSSW_7_4_X.

**Comment by ferencek** [comment]
In addition, it looks like several unrelated commits somehow got picked up in the process.

**Comment by gpetruc** [comment]
Hi Sal,

Let's start with having this info added to the jets (which only needs
changes to the producer that is now adding sub-jets), since we need it by
the end of this week for 7.4.0

We can then later discuss whether and how we need to provide the
convenience interface to get full list daughters, as that doesn't change
the persistent dataformat and so it can be added in a later 7.4.x release
and so we have time until the first samples really start to appear.

Giovanni

Il 10-mar-2015 00:38 "rappoccio" <notifications@github.com> ha scritto:
>
> @gpetruc It's basically a schema shift either way. The number of people
who will recluster AFTER this pull request will be much fewer than those
before because this is the first time subjet b-tagging is usable, and
basically everyone except W/Z tagging needs it.
>
> Nevertheless, if you guys desperately want this, I can try to implement
it tomorrow (although I think it's better in pat::Jet than in
CompositeCandidatePtr).

**Comment by deguio** [comment]
+1
doesn't affect DQM

**Comment by gpetruc** [comment]
Hi Sal,

An implementation of the re-adding of the daughters is in https://github.com/gpetruc/cmssw/commit/e406ed27ae3502fd8a7c82e1d2cafb167d168727

**Comment by rappoccio** [comment]
@gpetruc Thanks a lot, I had come up with something similar (but wasn't debugged yet). 

I'll put in the PR. This PR got completely mangled by my git rebase anyway so I will close it and reopen.
